src/common/container.h

```diff
@@ -80,61 +81,99 @@ class FixedArray {
   T data_[N];
 };
 
-// A simple fixed size, compile-time allocated deque.
-// Any positiv CAPACITY parameter above 1 should work. Choosing a power of
-// two will typically generate faster code as modulo ops can become bit-AND ops.
-//
-// For a cheap empty() and size() implementation, the actual capacity() is one
-// less than the chosen CAPACITY template parameter.
-template <typename T, int CAPACITY>
-class RingDeque {
- public:
-  RingDeque() {
-    static_assert(CAPACITY > 1, "Capacity needs to be > 1");
-    // intentionally not initializing memory to better see if users do.
+template <typename T, typename... Args>
+std::unique_ptr<T> make_unique(Args &&...args) {
+  return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
+}
+
+template <typename T, size_t CAPACITY>
+struct FixedStoragePolicy {
+  static_assert(CAPACITY > 1, "Capacity needs to be > 1");
+  // intentionally not initializing memory to better see if users do.
+  T buffer[CAPACITY];
+  using type = T;
+  T *GetBuffer() { return &buffer[0]; }
+  static constexpr size_t storage_size = CAPACITY;
+};
+
+template <typename T>
+struct DynamicStoragePolicy {
+  explicit DynamicStoragePolicy(const size_t size)
+      : storage_size(size), buffer(std::unique_ptr<T[]>(new T[size])) {
+    assert(size > 1 && "Capacity needs to be > 1");
   }
 
-  size_t size() const { return (write_pos_ + CAPACITY - read_pos_) % CAPACITY; }
+  using type = T;
+  size_t storage_size;
+  std::unique_ptr<T[]> buffer;
+  T *GetBuffer() { return buffer.get(); }
+};
+
+template <typename StoragePolicy>
+class RingDeque : private StoragePolicy {
```

Comment:
For clarity, ideally, we should use composition, not inheritance.
I suppose you attempted it without, but then ran into the problem that it requires a constructor with size for the runtime container ?

Reply:
I think it's not an issue to go for composition, it's just more boilerplate to add. With inheritance things are fairly more straightforward although a bit less flexible. Let me know your preference!